sklearn/cluster/tests/test_k_means.py

```diff
@@ -787,66 +743,61 @@ def test_k_means_init_fitted_centers(data):
     assert_array_almost_equal(centers, new_centers)
 
 
-def test_less_centers_than_unique_points():
+def test_kmeans_warns_less_centers_than_unique_points():
+    # Check KMeans when the number of found clusters is smaller than expected
     X = np.asarray([[0, 0],
                     [0, 1],
                     [1, 0],
                     [1, 0]])  # last point is duplicated
+    km = KMeans(n_clusters=4)
 
-    km = KMeans(n_clusters=4).fit(X)
-
-    # only three distinct points, so only three clusters
-    # can have points assigned to them
-    assert set(km.labels_) == set(range(3))
-
-    # k_means should warn that fewer labels than cluster
-    # centers have been used
-    msg = ("Number of distinct clusters (3) found smaller than "
-           "n_clusters (4). Possibly due to duplicate points in X.")
-    assert_warns_message(ConvergenceWarning, msg, k_means, X,
-                         sample_weight=None, n_clusters=4)
+    # KMeans should warn that fewer labels than cluster centers have been used
+    msg = (r"Number of distinct clusters \(3\) found smaller than "
+           r"n_clusters \(4\). Possibly due to duplicate points in X.")
+    with pytest.warns(ConvergenceWarning, match=msg):
+        km.fit(X)
+        # only three distinct points, so only three clusters
+        # can have points assigned to them
+        assert set(km.labels_) == set(range(3))
 
 
 def _sort_centers(centers):
     return np.sort(centers, axis=0)
 
 
 def test_weighted_vs_repeated():
```

Comment:
This can't be guaranteed in most situations. It's only by chance that the test currently passes. Changing seed or n_init makes it fail. This is due to rng in sampling X, for the init or for the minibatches.

I changed the test to only cover the guaranteed situation: KMeans with precomputed init.